### PR TITLE
EOSPAC vector implementations

### PR DESCRIPTION
**Pull request by rbberger**
## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file

### CODE REVIEW

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_eospac.hpp:312]: Honestly there's not much we can do when the heat capacity (DEDT) is zero. If it's negative then the EOS is fundamentally flawed.

I'd be in favor of an assert here so that if somebody gets crazy results they can run in debug mode and see that their EOS has serious issues, but since this code was here before your modifications I'll create an issue for this so we can handle it later.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_eospac.hpp:118]: We should probably have a way to compute the required scratch space and hand that to the user. I'm not sure the best way to do this though.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_eospac.hpp:76]: I assume that if you want to use the `CHUNK_SIZE` template parameter, you need to use the `EOSPAC` class directly since this isn't exposed through the `Variant` class right? I don't think I'd advocate for any changes to the `Variant` class to accommodate this yet since it's a requirement specific to EOSPAC and unlikely to be needed by other EOS.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_eospac.hpp:596]: I think we should get rid of this `max` but again since this was here before, I don't think it needs to be changed in this MR. Stuff like this should be handled in a production hardening layer I think.

**Review by jhp-lanl — CHANGES_REQUESTED** [review]
I think this looks really good! I don't think the underlying math and thermodynamics you copied over from the existing code has been thoroughly vetted, but that's beyond the scope of this MR. There may be additional performance improvements we can add, but those can be done later.

The only changes I would ask for would be a simple function to provide a required scratch size and a simple test. There are multiple scratch sizes needed though so this requires some discussion I think. For the test, I would just pick one of these methods and make a test for the chunk version and the full array version and make sure the result agrees with a for loop over scalar lookups. We can expand the test later as part of vetting the underlying logic and math.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:118]: Actually, can't we just use the `lambdas` variable for scratch? EOSPAC currently says it doesn't need any lambdas, but we could set that requirement to 8 to account for the largest possible scratch usage.

In fact, we could also overload `nlambdas()` to take an integer corresponding to the number of points to ease the calculation of lambda arrays for spiner as well.

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp:76]: Yes, that's the intention. It's a tradeoff between using more on-demand stack vs preallocating a huge heap scratch space based of `nscratch() * nelements * sizeof(Real)`. Which of the two implementations or if both should stay is still up for debate. I can imagine use cases for either one of them.

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp:118]: The required scratch space is `nscratch() * elements * sizeof(Real)`. We could have a helper function that just takes `nelements` and returns the number of bytes.

Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp:118]: One possible solution for returning the exact scratch space requirement per call is having a function that returns the right value for each member function pointer.

File: singularity-eos/eos/eos_eospac.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_eospac.hpp:106]: Have you tested the chunk size version of the code on GPUs? I'm pretty sure as written, this will not work, as the chunk are filled on CPU, whereas the data may live on GPU. This might be fixable with synchronized memory or with explicit deep copies, but it's something to be aware/concerned about.

File: singularity-eos/eos/eos_eospac.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_eospac.hpp:195]: It's pretty painful how much duplicated code lives in `FillEos`. I wonder if there's a way to reduce this.

File: singularity-eos/eos/eos_eospac.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_eospac.hpp:312]: I think this probably merits a larger discussion.

**Review by Yurlungur — COMMENTED** [review]
This is a draft, so not approving or requesting changes. A little bit of feedback below.

**Comment by Yurlungur** [comment]
Also thanks for putting in the work to prototype this! I like both the "scatch" and "chunk" approaches. But I think we should be a little thoughtful of GPU performance in both cases. I don't think any of this code will run on GPU as-is.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:597]: Again like you said, larger discussion. But one last comment and I'll leave it to another time.

My changes to the Gruneisen EOS were mainly targeted at sanitizing inputs to produce realistic outputs. Since there's no way to _a priori_ know what inputs will lead to bad inputs for sesame tables, I feel that it's not necessarily our place to modify what the EOS returns _without the user explicitly asking for it_. That's why I argue for a production hardening modifier to handle some of these things. The behavior of the modifier can be documented so that the user knows what they are guaranteed to be handed. I feel that without asking for a modifier like this, the user should be guaranteed the best representation of the EOS itself even if that data is garbage.

But again, this needs a larger discussion beyond the scope of this MR.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:313]: Discussion: https://github.com/lanl/singularity-eos/issues/180

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:196]: Could the chunk version call the regular version but with chunked data?

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp:123]: @Yurlungur Added a `max_scratch_size(nelements)` method

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp]: Updated all loops to portableFor in those cases.

**Comment by rbberger** [comment]
@Yurlungur @jhp-lanl I've updated all scratch and chunk versions to use EOSPAC options to minimize copies and do unit conversions. I've also expanded the profiling to include more methods and made it part of CTest. Below are timings I've generated per method on `skylake-platinum` nodes on Darwin:
EOSPAC_vectorization_time

EOSPAC_vectorization_throughput

**Comment by jhp-lanl** [comment]
Great work @rbberger ! I'm excited especially that this work is part of `CTest` so that we can have a baseline and look for ways to improve. We should discuss this more at the meeting tomorrow!

**Comment by rbberger** [comment]
@jhp-lanl yeah, having it part of CTest already pays off. I'm having weird failures on power9 that need further investigation. :smile:

**Comment by Yurlungur** [comment]
Nice work @rbberger ! I second @jhp-lanl 's comments.

**Comment by Yurlungur** [comment]
Peanut gallery says we go to scratch and drop lambdas and chunks

**Comment by rbberger** [comment]
@Yurlungur @jhp-lanl Issue #187 blocks me from doing any meaningful GPU tests on the CI.

**Comment by jhp-lanl** [comment]
> @Yurlungur @jhp-lanl Issue #187 blocks me from doing any meaningful GPU tests on the CI.

I seem to recall that the accessible EOSPAC spackage isn't up-to-date and I'm not quite sure what the best way to get 6.5 is. I'll ping Daniel Sheppard.

**Comment by rbberger** [comment]
@Yurlungur ready for review.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp]: Since we decided to punt on getting EOSPAC working on GPUs, I'm resolving

File: python/module.cpp
Comment (jhp-lanl) [inline python/module.cpp:116]: I assume it's the `use_scratch` template parameter below, right? Would it make sense to add a comment or just introduce a `bool eospac_use_scratch = true` variable to explain this since it deviates from the construction of the other classes?

File: python/module.hpp
Comment (jhp-lanl) [inline python/module.hpp:85]: Is the "WithScratch" variant of the functions tested somewhere?

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_eospac.hpp:418]: Just a comment for the future for all these derivative quantities that require two lookups: we'll want to work with the EOSPAC team to find more performant ways to do this. If you _just_ want the bulk modulus and nothing else, we can probably accomplish this with a single lookup and let EOSPAC handle the thermodynamics to figure out how to do it.

Again... no changes needed as this is a future consideration and I'll make an issue.

File: test/CMakeLists.txt
Comment (jhp-lanl) [inline test/CMakeLists.txt:110]: Thanks again for adding this profiling test! This is excellent and we need to do more of this!

File: test/profile_eospac.cpp
Comment (jhp-lanl) [inline test/profile_eospac.cpp:79]: Were these for debugging or do we need to leave them in?

File: test/CMakeLists.txt
Comment (jhp-lanl) [inline test/CMakeLists.txt:88]: This may be my own ignorance, but do we need to exempt CUDA or can it operate on the non-Kokkos code for the test?

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_eospac.hpp:724]: I like this!

**Review by jhp-lanl — APPROVED** [review]
I'm nominally approving this because most of my comments are questions rather than critiques and I can be convinced out of any code changes.

@rbberger feel free to resolve any of my questions as you answer them or make code changes to address them.

File: singularity-eos/eos/eos_eospac.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_eospac.hpp:722]: Since the scratch methods are available in the variant, should the scratch size methods be available too? We can implement the "null" versions in the base class and just return 0?

File: test/CMakeLists.txt
Comment (Yurlungur) [inline test/CMakeLists.txt:111]: This is clever.

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Sorry I left this so long @rbberger it slipped my mind. Overall this looks very good. A few nitpicks below.

File: singularity-eos/eos/eos_base.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_base.hpp:84]: In C++ 17 I could use `[[maybe_unused]]` in this context to still give it a name, while silencing the compiler warning. As a compromise I could do `/* scratch */` for now.

File: test/profile_eospac.cpp
Comment (rbberger) [reply test/profile_eospac.cpp:79]: It was left as a comment to document how to test different methods. The CMake logic generates executables with each one of these defines. I have no preferences either way.

File: test/CMakeLists.txt
Comment (rbberger) [reply test/CMakeLists.txt:90]: I remember I had to do this because it was failing the regit CI otherwise.

File: test/profile_eospac.cpp
Comment (jhp-lanl) [reply test/profile_eospac.cpp:79]: We can leave them, but can you add a comment above to just explain what's going on here?

File: singularity-eos/eos/eos_base.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:84]: I'll defer to your judgement @rbberger

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:722]: I remember discussing this before and I thought we decided that it was specific enough to EOSPAC that a host code would need to differentiate anyway. However, I'm less convinced now that that's the right way to go, especially if we use EOSPAC in the PTE solvers. I'd support including null versions in the variant.

File: python/module.cpp
Comment (rbberger) [reply python/module.cpp]: Fixed.

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp]: Removed the ones that weren't used.

File: test/profile_eospac.cpp
Comment (rbberger) [reply test/profile_eospac.cpp:83]: Added a comment above.

File: eospac-wrapper/eospac_wrapper.cpp
Comment (rbberger) [reply eospac-wrapper/eospac_wrapper.cpp:209]: Didn't measure that, but my guess it is smaller than setting ALL options to a defined state before each call, which would be necessary if you don't reset them right away. Otherwise, you can't rely on the option values, given that you don't know what other EOSPAC calls have happened before each function.

Comment (rbberger) [reply eospac-wrapper/eospac_wrapper.cpp:209]: While the table handle is associated with the EOSPAC class, how do we know im which order the available functions are called? At a minimum we need to set all options that potentially could have been set by any member function to avoid interference.

Comment (rbberger) [reply eospac-wrapper/eospac_wrapper.cpp:209]: Ok, as an example: if you look at the usage of the `PofRT_table_`, the calls happen sometimes with `F_CONVERT` and sometimes with `X_Y_PASSTHROUGH` options passed. I can not predict in what order the different EOS functions are called, so I have to make sure the options are in a defined state.

Comment (Yurlungur) [reply eospac-wrapper/eospac_wrapper.cpp:209]: Ok I see. Let's leave this for now then. Maybe worth putting a TODO to talk with the EOSPAC folks about though. If setting options implies a performance hit we might want to redesign the class to avoid having to do so.

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_base.hpp]: Let's do `/* scratch */` as we can't get `C++17` yet.

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp]: Done.

File: singularity-eos/eos/eos_base.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_base.hpp]: This is already done.

File: singularity-eos/eos/eos_ideal.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_ideal.hpp:112]: These "0 size" funcs could probably go in the base class. Either way though. :)

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_variant.hpp:291]: Should the non-scratch vector calls be removed? Or perhaps just default scratch to a nullptr?

**Review by Yurlungur — APPROVED** [review]
I approve and am ready to merge, pending the items on the checklist being checked off.

A few things on my mind. But we don't have to resolve these now, we can merge first and tackle in a second PR.
- Should we have a correctness test? Perhaps use the vector calls and the scalar calls for the EOSPAC backend and compare? Or alternatively compare EOSPAC and spiner, as is currently done, but convert the EOSPAC calls to vector calls? We could make it work for CPU I imagine?
- Still need to get GPU working, but that's for later as discussed previously.
- Make sure to trigger tests on Darwin and make sure they pass. :)
- Make sure to do changelog, CC, etc.
- Perhaps the scratch API and the particulars of the EOSPAC passthrough should be documented
- We'll have to figure out how to do the passthrough for modifiers... I think something is needed there. Probably we can leverage the scratch machinery.

File: singularity-eos/eos/eos_variant.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_variant.hpp:291]: Doesn't the python interface benefit from the non-scratch vector calls?

File: singularity-eos/eos/eos_eospac.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_eospac.hpp]: I opted for removing the template parameters and just forcing pointer types. A `static_assert` or SFINAE with `enable_if_t` would have been a bit over-engineered in my opinion.

File: singularity-eos/eos/eos_ideal.hpp
Comment (rbberger) [reply singularity-eos/eos/eos_ideal.hpp:112]: I followed the patterns I saw in PreferredInput and other functions.

**Comment by rbberger** [comment]
> * Should we have a correctness test? Perhaps use the vector calls and the scalar calls for the EOSPAC backend and compare? Or alternatively compare EOSPAC and spiner, as is currently done, but convert the EOSPAC calls to vector calls? We could make it work for CPU I imagine?

Actually, the `profile_eospac` test does include a comparison with the scalar versions.

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_variant.hpp:291]: Ah good point. :+1:

File: python/module.hpp
Comment (rbberger) [reply python/module.hpp:85]: @jhp-lanl just added Python tests that do so: https://github.com/lanl/singularity-eos/blob/59a2e74f784d6b1ce36304331893a74f62512c89/test/python_bindings.py#L528

**Review by jhp-lanl — APPROVED** [review]
This looks really good @rbberger ! Thanks for all the effort on this!

**Comment by jhp-lanl** [comment]
I'll wait for the Darwin tests to pass before we merge.

**Review by Yurlungur — APPROVED** [review]
Looks good. :+1: Let's merge as soon as Darwin tests pass

**Comment by jhp-lanl** [comment]
@rbberger is there anything holding this up? To me it looks like you can merge in master (and resolve conflicts) and we can merge this.

**Comment by rbberger** [comment]
> @rbberger is there anything holding this up? To me it looks like you can merge in master (and resolve conflicts) and we can merge this.

No. It kind of got stuck because we were looking at why GPU builds being so slow (in general, not just this), but I'll fix this one up again so we can merge it.

**Comment by rbberger** [comment]
@Yurlungur re-git pipeline successful. feel free to merge.

**Comment by Yurlungur** [comment]
Thanks for all your hard work on this @rbberger ! At some point we should check to see what is necessary to pull this through for GPU implementations and modifiers. But for now I think we can celebrate speeding up the EOSPAC calls.

**Comment by jhp-lanl** [comment]
Yes indeed! Thanks for all the hard work @rbberger !

**Pull request merged**
